src/common.c

```diff
@@ -151,24 +151,34 @@ size_t hexlenl(const char *q, int *extra_chars)
 	return _hexlen(q, atoi16l, extra_chars);
 }
 
+static int isdec_len(const char *q, const char *mxv)
+{
+	const char *p = q;
+	do
+	{
+		if (*p < '0' || *p > '9' || p - q >= 10)
+			return 0;
+	} while (*++p);
+
+	return p - q < 10 || strcmp(q, mxv) <= 0;
+}
 int isdec(const char *q)
 {
-	char buf[24];
-	int x = atoi(q);
-	sprintf(buf, "%d", x);
-	return !strcmp(q,buf) && *q != '-';
+	const char *imax = "2147483647";
+	return isdec_len(q, imax);
 }
 int isdec_negok(const char *q)
 {
-	char buf[24];
-	int x = atoi(q);
-	sprintf(buf, "%d", x);
-	return !strcmp(q,buf);
+
+	const char *imx = "2147483648";
+	if (*q != '-')
+		return isdec(q);
+
+	return isdec_len(q + 1, imx);
```

Comment:
Also no need to define `imx`, and you can optionally simplify the whole function body to:

    return *q == '-' ? isdec_len(q + 1, "2147483648") : isdec(q);